### ISSUE TITLE
Proper handling of special characters in HTTP Cookies

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#15218]
This blog post does a deep dive into some intricasies of parsing cookies in both server and client applications: https://grayduck.mn/2024/11/21/handling-cookies-is-a-minefield/. Yeah the title expresses it very well: This is a minefiled 💣 

Part of the investigation compares how popular HTTP libraries handle cookie parsing. So naturally I went ahead to see how Crystal's `HTTP::Cookie` fares.
Seems like cookie creation via `HTTP::Cookie.new` follows [RFC 6235 (Section 4.1.1)](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-set-cookie) (*how servers should send cookies*) to the letter, raising if the value contains any special characters or Unicode characters.
But that means it disregards [RFC 6265 (Section 5.6)](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-the-set-cookie-header-field) (*how browsers should accept cookies*) which allows a number of special characters and doesn't have a concise opinion on Unicode.
`HTTP::Cookie` is intended for use in both client and server and does not distinguish between different modes.
In practice this means `HTTP::Cookie` is lacking when used in an HTTP client.

| Method | CTLs | htab | space | dquote | comma | backslash | delete | 0x80-FF(and Unicode) |
|-|-|-|-|-|-|-|-|-|
[RFC 6235 (Section 4.1.1)](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-set-cookie) | No | No | No | No | No | No | No | No
[RFC 6265 (Section 5.6)](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-the-set-cookie-header-field) | No | Yes | Yes | Yes | Yes | Yes | No | ??
| `HTTP::Cookie.new` | No | No | No | No | No | No | No | No |

The methods for parsing multiple cookies from an HTTP headers are `HTTP::Cookies.from_client_headers` and `HTTP::Cookies.from_server_headers`. They handle parsing errors, but do it differently:
* `from_client_headers` reads the value of the cookie up to the first invalid byte.
* `from_server_headers` drops the cookie pair entirely if the value contains an invalid byte (this is in accordance with [RFC 6265 Section 5.7](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-storage-model)).

```cr
HTTP::Cookies.from_client_headers(
  HTTP::Headers {"Cookie" => "ginger=snap; cookie=hm🍪delicious; snicker=doodle",}
).to_s # => "ginger=snap; cookie=hm; snicker=doodle"
HTTP::Cookies.from_server_headers(
  HTTP::Headers {"Set-Cookie" => ["ginger=snap", "cookie=hm🍪delicious", "snicker=doodle"],}
).to_s # => "ginger=snap; snicker=doodle"
``` 

What to do about all of this, I'm not entirely sure. But I suppose we should make cookie parsing more lenient. Especially client-side parsing is definitely too restrictive. And on the server side we should probably also try to avoid breaking cookie usage patterns that may violate the RFC but still are commonly used.

Related:  #10983

**Comment by ysbaddaden** [comment]
Server: Dropping the cookie ain't the worst (that would be raising an exception), and it's even conforming to the spec. I wouldn't make any change, but improve the documentation to state the behavior?

Client: We don't conform to the spec, so yeah, we should accept more characters. Reading up to the invalid codepoint is weird, though.

**Comment by bararchy** [comment]
> But I suppose we should make cookie parsing more lenient

Yes, we should basically allow everything that a browser will (Chrome in this case as the market leader). to ensure we can work with the same web "standards"

**Comment by straight-shoota** [comment]
I figure this could also be a good excuse to get rid of the [unspeakable complexity of a regular expression](https://github.com/crystal-lang/crystal/blob/6928ca70e5e31c1481a308f07818c296ca005a0f/src/http/cookie.cr#L197-L228) which currently handles cookie parsing.

**Comment by bararchy** [comment]
@straight-shoota I think this issue can be closed after #15244 and #15500 got handled no?

**Comment by Sija** [comment]
@bararchy The issue was not resolved in any way though, just documented.
